kivy/uix/widget.py

```diff
@@ -379,6 +396,78 @@ def __hash__(self):
     def __self__(self):
         return self
 
+    def apply_class_lang_rules(
+            self, root=None, ignored_consts=set(), rule_children=None):
+        '''
+        Method that is called by kivy to apply the kv rules of this widget's
+        class.
+
+        :Parameters:
+            `root`: :class:`Widget`
+                The root widget that instantiated this widget in kv, if the
+                widget was instantiated in kv, otherwise ``None``.
+            `ignored_consts`: set
+                (internal) See :meth:`~kivy.lang.builder.BuilderBase.apply`.
+            `rule_children`: list
+                (internal) See :meth:`~kivy.lang.builder.BuilderBase.apply`.
+
+        This is useful to be able to execute code before/after the class kv
+        rules are applied to the widget. E.g. if the kv code requires some
+        properties to be initialized before it is used in a binding rule.
+        If overwriting remember to call ``super``, otherwise the kv rules will
+        not be applied.
+
+        In the following example,
+
+        .. code-block:: python
+
+            class MyWidget(Widget):
+                pass
+
+            class OtherWidget(MyWidget):
+                pass
+
+        .. code-block:: kv
+
+        <MyWidget>:
+            my_prop: some_value
+
+        <OtherWidget>:
+            other_prop: some_value
+
+        When ``OtherWidget`` is instantiated with ``OtherWidget()``, the
+        widget's :meth:`apply_class_lang_rules` is called and it applies the
+        kv rules of this class - ``<MyWidget>`` and ``<OtherWidget>``.
+
+        Similarly, when the widget is instantiated from kv, e.g.
+
+        .. code-block:: kv
+
+            <MyBox@BoxLayout>:
+                height: 55
+                OtherWidget:
+                    width: 124
+
+        ``OtherWidget``'s :meth:`apply_class_lang_rules` is called and it
+        applies the kv rules of this class - ``<MyWidget>`` and
+        ``<OtherWidget>``.
+
+        .. note::
+
+            It applies only the class rules not the instance rules. I.e. in the
+            above kv example in the ``MyBox`` rule when ``OtherWidget`` is
+            instantiated, its :meth:`apply_class_lang_rules` applies the
+            ``<MyWidget>`` and ``<OtherWidget>`` rules to it - it does not
+            apply the ``width: 124`` rule. The ``width: 124`` rule is part of
+            the ``MyBox`` rule and is applied by the ``MyBox``'s instance's
+            :meth:`apply_class_lang_rules`.
+
+        .. versionchanged:: 1.11.0
+        '''
+        Builder.apply(
+            self, ignored_consts=ignored_consts,
+            rule_children=rule_children)
```

Comment:
I feel uneasy about passing the rule_children in all these methods. Could it be returned instead, and thus accumulated in the calls up the call stack? *This is just a guess, I have not analyzed the code thoroughly.

Reply:
It could be returned as you suggest, but I find it worse to return that from apply, rather than appending to a rule_children list passed. 

The reason is that `apply` applies kv rules and if it were to return something it should be the root widget, or maybe the list of rules applied or something similar. I.e. it should return something kv rule related that was just applied - and in the future that's how we might change it so that it returns something like that. 

If we were to return a list of children, that would preclude being able to change it to return something more relevant in the future. And specifically, returning a list of children doesn't seem relevant, except for this specific need to dispatch `on_kv_post`. If we didn't have this need, we would never return the list of children as this doesn't seem useful elsewhere in how that method would be used. That's why it seemed more appropriate to pass the list of children to be filled.

Can you explain why you feel uneasy? I get that it's not particularity pretty, but I don't see issues with it. Function parameters we have many - we have but one return statement.